node/network/approval-distribution/src/lib.rs

```diff
@@ -265,47 +429,111 @@ struct BlockEntry {
 	candidates: Vec<CandidateEntry>,
 	/// The session index of this block.
 	session: SessionIndex,
+	/// Approval entries for whole block. These also contain all approvals in the case of multiple candidates
+	/// being claimed by assignments.
+	approval_entries: HashMap<(ValidatorIndex, CandidateBitfield), ApprovalEntry>,
```

Comment:
still wrapping my head around this mapping and whether it can be simplified

Reply:
Yes, I also hate this. I could remove it, but importing approvals would be less efficient. We'll have to iterate all the `approval_entries` and check candidate `CandidateBitfield` if it has the `candidate_index` for each approval in `import_and_circulate_approval`. This would scale really bad with `n_validators`, while the current implementation always needs 2 hashmap lookups - see `get_approval_entry`.

Reply:
Simply put, to get rid of the extra mapping `CandidateIndex -> (ValidatorIndex, CandidateBitfield)` we'd need a `HashMap<(ValidatorIndex, CandidateBitfield), ApprovalEntry>,` which provides a a getter for `(ValidatorIndex, CandidateIndex)`.

Reply:
to be clear, are keys unique by `ValidatorIndex`?

Reply:
They are not unique, there could be different assignments for a `ValidatorIndex` which mean a different candidate bitfield as well.